### ISSUE TITLE
WASM support

### ISSUE THREAD

**Issue by JakeWharton** [drewhamilton/Poko#235]
I tried adding WASM support quickly now that AssertK released with it, but the test runs fail. Not sure why. Output is pretty useless.
```
> Task :poko-tests:wasmNodeTest FAILED
Caching disabled for task ':poko-tests:wasmNodeTest' because:
  Build cache is disabled
Task ':poko-tests:wasmNodeTest' is not up-to-date because:
  Task is untracked because: Should always re-run for WASM
Starting process 'command '/Users/jw/.gradle/nodejs/node-v20.0.0-darwin-arm64/bin/node''. Working directory: /Volumes/dev/other/Poko/build/js/packages/Poko-poko-tests-wasm-test Command: /Users/jw/.gradle/nodejs/node-v20.0.0-darwin-arm64/bin/node --experimental-wasm-gc /Volumes/dev/other/Poko/build/js/packages/Poko-poko-tests-wasm-test/kotlin/runUnitTests.mjs --dryRun
Successfully started process 'command '/Users/jw/.gradle/nodejs/node-v20.0.0-darwin-arm64/bin/node''
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]

FAILURE: Build failed with an exception.
```
Will try again later, but maybe someone else wants to look in the mean time.

Branch here: https://github.com/drewhamilton/Poko/compare/main...JakeWharton:Poko:jw.wasm.2023-09-14

**Comment by drewhamilton** [comment]
> ```
> Working directory: /Volumes/dev/other/Poko/build/js/packages/Poko-poko-tests-wasm-test Command: /Users/jw/.gradle/nodejs/node-v20.0.0-darwin-arm64/bin/node --experimental-wasm-gc /Volumes/dev/other/Poko/build/js/packages/Poko-poko-tests-wasm-test/kotlin/runUnitTests.mjs --dryRun
> ```

Running this directly gives me more helpful error information. Looks like it's a Poko problem:
```
file:///.../Poko/build/js/packages/Poko-poko-tests-wasm-test/kotlin/Poko-poko-tests-wasm-test.uninstantiated.mjs:199
        const wasmModule = new WebAssembly.Module(wasmBuffer);
                           ^

CompileError: WebAssembly.Module(): Compiling function #5957:"poko.AnyArrayHolder.equals" failed: call[0] expected type (ref null 1475), found struct.get of type (ref null 146) @+484610
    at instantiate (file:///.../Poko/build/js/packages/Poko-poko-tests-wasm-test/kotlin/Poko-poko-tests-wasm-test.uninstantiated.mjs:199:28)
    at async file:///.../Poko/build/js/packages/Poko-poko-tests-wasm-test/kotlin/Poko-poko-tests-wasm-test.mjs:3:17

Node.js v20.0.0
```

**Comment by JakeWharton** [comment]
Ah, nice.

**Comment by drewhamilton** [comment]
Narrowed down the issue to `@ArrayContentBased` on generic and `Any` properties, i.e. the runtime array determination logic. If I comment those 3 classes + tests out, the rest pass.

**Comment by drewhamilton** [comment]
After rebasing on Kotlin 1.9.21, I get a new error.

```
> Task :poko-tests:wasmJsNodeTest FAILED

Please make sure that your runtime environment supports the latest version of Wasm GC and Exception-Handling proposals.
For more information, see https://kotl.in/wasm-help

file:///.../Poko/build/js/packages/Poko-poko-tests-wasm-js-test/kotlin/Poko-poko-tests-wasm-js-test.uninstantiated.mjs:195
        const wasmModule = new WebAssembly.Module(wasmBuffer);
                           ^

CompileError: WebAssembly.Module(): invalid value type 0x71 @+364
    at instantiate (file:///.../Poko/build/js/packages/Poko-poko-tests-wasm-js-test/kotlin/Poko-poko-tests-wasm-js-test.uninstantiated.mjs:195:28)
    at async file:///.../Poko/build/js/packages/Poko-poko-tests-wasm-js-test/kotlin/Poko-poko-tests-wasm-js-test.mjs:3:17

Node.js v21.0.0
```

Same thing happens on Node.js v21.4.0. Same thing happens if I comment out all the code in `:poko-tests` and comment out the addition of the Poko compiler plugin to `:poko-tests`, so it's not Poko-specific.

Looks like an exact match for [this issue](https://youtrack.jetbrains.com/issue/KT-63014/Running-tests-with-wasmJs-in-1.9.20-requires-Chrome-Canary), but I still get the same error after bumping to Node.js 21.0.0 and confirming my default browser is 119+.

[This comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1845374#c3) suggests `0x71` is a null reference, so maybe `wasmBuffer` is null on the quoted JS line? 🤷‍♂️

I pushed a [drew/wasm](https://github.com/drewhamilton/Poko/compare/main...drew/wasm) branch, which just rebases your branc, updates the `wasm()` invocations to `wasmJs()` invocations, and bumps to Node.js 21. Would love to know if you can glean any more information about this error, otherwise I can file a YouTrack issue.

**Comment by drewhamilton** [comment]
Ah, I assumed the canary listed in the YouTrack issue would have landed in the since-released stable 21.0.0. Thanks!

Now I'm back to the original error, but with a nicer message this time:
```
CompileError: WebAssembly.Module(): Compiling function #5970:"poko.AnyArrayHolder.equals" failed: call[0] expected type (ref null 1466), found struct.get of type (ref null 146) @+482088
    at instantiate (file:///.../Poko/build/js/packages/Poko-poko-tests-wasm-js-test/kotlin/Poko-poko-tests-wasm-js-test.uninstantiated.mjs:195:28)
    at async file:///.../Poko/build/js/packages/Poko-poko-tests-wasm-js-test/kotlin/Poko-poko-tests-wasm-js-test.mjs:3:17
```

**Comment by JakeWharton** [comment]
Yeah I've been waiting for a normal Node to land as well, but seems like it's not happening any time soon, unfortunately.
